### DISCUSSION TITLE
How to reduce the memory required by Netket？

### DISCUSSION THREAD

**Discussion by Kaho-shur** [netket/netket#2146] [Q&A] [answered]
Dear Developers,  

When using NetKet to process a 32×32 system, we encounter **GPU memory overflow** . So far, we have tried the following optimizations to reduce memory usage:  
- Lowering `n_sample`, `chunk_size`, `sweep_size`, and `n_chains_per_rank`;  
- Adjusting the optimization method (switching from Stochastic Reconfiguration (SR) to `solver=cg`);  
- Reducing data precision (from `float64` to `float32`).  

However, the memory overflow issue still persists (see Figure 1). Are there any other parameter adjustments or optimization strategies we can adopt to resolve this problem?  

<img width="1680" height="183" alt="image" src="https://github.com/user-attachments/assets/889635cb-047f-46bd-82e5-e5056d39be4b" />

**Comment by PhilipVinc** [comment]
What driver and how are using ? What options? What is the model?

**Reply by Kaho-shur** [reply]
We use a 3-layer CNN neural network (with the number of features/neurons in each layer being 32, 64, and 128 respectively) as the trial wavefunction for Variational Monte Carlo (VMC). The detailed code is as follows:
```
    sa = nk.sampler.MetropolisExchange(
    hilbert_space, sweep_size=sweep_size, graph=lattice,
    n_chains_per_rank=2      
    )
```
`model = CNN2D(hidden_layers=tuple(hidden_layers), Lx=Lx, Ly=Lx)`
```
class CNN2D(nn.Module):
    Lx: int                  
    Ly: int | None = None    
    hidden_layers: tuple[int, ...] = (32, 64, 128)
    kernel_size: int = 2     
    dense_layer: int = 16

    @nn.compact
    def __call__(self, x: jnp.ndarray) -> jnp.ndarray:

        Ly = self.Ly or self.Lx

        if x.ndim == 1:
            x = x[None, :] 
        B, N = x.shape
        assert N == self.Lx * Ly, f"Input length {N} != Lx*Ly = {self.Lx*Ly}"

        x = x.reshape(B, self.Lx, Ly, 1)

        k = (self.kernel_size, self.kernel_size)
        for h in self.hidden_layers:
            x = nn.Conv(features=h, kernel_size=k, padding="SAME")(x)
            x = nn.relu(x)

        x = x.reshape(B, -1)          
        x = nn.Dense(self.dense_layer)(x)
        x = nn.relu(x)
        y= jnp.sum(x, axis=-1)
        return y   
```
```
    optimizer = nk.optimizer.Adam(learning_rate=lr)
    gs = nk.driver.VMC(
        H, optimizer, variational_state=vstate,
        preconditioner=nk.optimizer.SR(diag_shift=0.1,solver=cg))
```

**Reply by PhilipVinc** [reply]
Are you running on… 1 cpu? Many CPUs? 1 gpu? Many?

**Reply by PhilipVinc** [reply]
How many parameters do you have?

**Reply by Kaho-shur** [reply]
One GPU, with the model (NVIDIA) A6000

**Reply by Kaho-shur** [reply]
Total number of parameters: 2,138,514

**Reply by PhilipVinc** [reply]
In any case, I suggest you use nkx.driver.VMC_SR (https://netket.readthedocs.io/en/stable/api/_generated/experimental/driver/netket.experimental.driver.VMC_SR.html ) instead of the driver you are using, as it has a few extras, notably switching to minSR/ntk when you have more parameters than samples.

the defaults should work well, but you should set chunk_size_bwd=something , generally lower than the chunk size in the variational state used for the forward pass only.

to find the good chunk size, I usually loop though many values halving it each time until it works.

first I test vstate.sample() to check that sampling works and if it does not I change the chunk size in the sampler , then vstate.expect(Hamiltonian) and if it fails I change the chunk size in the variational state, then driver.step(1) and if it fails I change chunk size bwd .

if something does not fit in memory, which operation is it?

**Reply by PhilipVinc** [reply]
Yeah, 2M parameters can only be trained on the ntk formulation of SR (minSR), otherwise you’ll be attempting to invert a 2M^2 dense matrix which is virtually impossible. With the ntk you’ll be inverting a number of samples ^2 matrix, so your cost will scale rapidly if you use a lot of samples. In general that will be your major bottleneck that chunking cannot control.
Start with 2k samples at most, and only increase later on.

so you need the VMC sr driver with use_ntk=True. 
It should default to that automatically in your case .
In general ‘on_the_fly=True’ (the default) will also be faster and require a bit less memory. Experiment to know for sure.

Important thing: I see from above that you are using 2 chains? That s very inefficient on GPUS, as they are natively parallel and essentially doing 1 Mcmc step for 1 or 100 chains has the same cost.
We typically set nchains >= 512 per gpu. Often nchains = nsamples //2 .
Also n discard per chain = 1 or 2.

this means your MCMC sampling will not be good. This is very visible at the beginning of your optimization, but I you will be much faster.

**Reply by gcarleo** [reply]
also please consider not using Adam and SR together, it's not well defined / justified ! Sgd + SR Is the way to go

**Reply by Kaho-shur** [reply]
> chunk_size_bwd
[sample=256.txt](https://github.com/user-attachments/files/22571848/sample.256.txt)
[sample=64.txt](https://github.com/user-attachments/files/22571849/sample.64.txt)
[chunk_size=1.txt](https://github.com/user-attachments/files/22571850/chunk_size.1.txt)


According to the error message, the problem should be occurring in `vstate.sample()`. I set `nsamples = 256` (which works normally in the 16×16 system), but even after reducing `nsamples` to 64, the same issue still persists. I also tried setting `chunk_size = 1`, but it still fails to run. The detailed error report is in the attachment.  

Thank you for mentioning that increasing the number of chains will speed up the operation. After resolving the memory issue, I will set the number of chains to `nsamples // 2`. The reason I currently set it to 2 is due to insufficient memory.

Now I will continue to try using `nkx.driver.VMC_SR`, together with the combination of SGD (Stochastic Gradient Descent) and SR. Thank you for your kind assistance.

**Reply by PhilipVinc** [reply]
The error you are seeing is not a memory error. It’s a crash in CUDA which might signal either a broken installation or a bug in Jax. I suggest you open a bug report with a script to reproduce on Jax’s repository

**Reply by PhilipVinc** [reply]
If you have an older Jax version you can also try updating

**Reply by PhilipVinc** [reply, marked as answer]
see https://github.com/jax-ml/jax/issues/32168

**Reply by Kaho-shur** [reply]
Dear Developers,  

Hello. Thank you for your reply. I will keep track of this thread to see if I can find a corresponding solution! Thank you again!  

Additionally, I have tried the `nkx.driver.VMC_SR` you mentioned, but it does not perform as well as `nk.driver.VMC`. As shown in the figure, we can see that `nkx.driver.VMC_SR` converges significantly faster (and also takes much less time), but its minimum energy (for labels p=VMCSR1, VMCSR2) is higher than that of `nk.driver.VMC` (for labels p=VMC1, VMC2).  

<img width="2600" height="1508" alt="density_compare" src="https://github.com/user-attachments/assets/8ea59ffa-28ae-4e0d-9f50-f174c2b818f3" />

My configurations are as follows:  
```
    gs = nk.driver.VMC(
        H, optimizer, variational_state=vstate,
        preconditioner=nk.optimizer.SR(diag_shift=0.1, solver=cg))
```
`    gs = nkx.driver.VMC_SR(hamiltonian=H, diag_shift=0.1, optimizer=optimizer, variational_state=vstate)  `


For the labels in the figure (such as hl32×64), they refer to the number of channels in the CNN: "hl32×64" means 32 channels in the first layer and 64 channels in the second layer of the CNN, and the rest follow the same convention.

**Reply by PhilipVinc** [reply]
VMC_SR defaults to `nk.optimizer.solver.cholesky` as a solver, while you are using `cg` for VMC. If you want to do a proper comparison, you should use the same with both and initialise the state with the same seed.
Please read the documentation of VMC_SR.

Cholesky is sometimes unstable, and a generally better choice, albeit more expensive, is `nk.optimizer.solver.pinv_smooth`. `cg` is generally considered less good than anything else, so your results are kind of striking.

Apart from your simulations that explode, where the choice of diag shift and learning rate are clearly wrong, you will have to tune diag shift and learning rate a bit.
VMCSR and VMC are mathematically equivalent, but they compute things differently so they will feel numerical noise differently. 

For example a diag shift of 0.1 is quite large and won't lead in general to accurate solutions. a better choice would be more in the range of 1e-3 or smaller. You will have to adjust the learning rate accordingly.
setting momentum to ~0.8 or 0.9 (and further reducing the learning rate) might also help.

**Reply by Kaho-shur** [reply]
Thank you for your suggestions! I will lower the diag_shift value, add momentum, and use the same solver to run the experiment again. Thank you again!
